python/ray/data/impl/lazy_block_list.py

```diff
@@ -25,122 +32,331 @@ class LazyBlockList(BlockList):
+        metadata = read_progress_bar.fetch_until_complete(list(unique_meta_refs))
+        ref_to_data = {
+            meta_ref: data for meta_ref, data in zip(unique_meta_refs, metadata)
+        }
+        metadata = [ref_to_data[meta_ref] for meta_ref in meta_refs]
+        self._fetched_metadata = metadata
+        return blocks, metadata
+
+    def compute_first_block(self):
+        """Kick off computation for the first block in the list.
+
+        This is useful if looking to support rapid lightweight interaction with a small
+        amount of the dataset.
+        """
+        if self._tasks:
+            self._get_or_compute(0)
+
+    def ensure_metadata_for_first_block(self) -> Optional[BlockMetadata]:
+        """Ensure that the metadata is fetched and set for the first block.
+
+        This will only block execution in order to fetch the post-read metadata for the
+        first block if the pre-read metadata for the first block has no schema.
+
+        Returns:
+            None if the block list is empty, the metadata for the first block otherwise.
+        """
+        if not self._tasks:
+            return None
+        metadata = self._tasks[0].get_metadata()
+        if metadata.schema is not None:
+            return metadata
+        try:
+            _, metadata_ref = next(self._iter_block_partition_refs())
+        except (StopIteration, ValueError):
+            # Dataset is empty (no blocks) or was manually cleared.
+            pass
+        else:
+            # This blocks until the underlying read task is finished.
+            metadata = ray.get(metadata_ref)
+            self._fetched_metadata[0] = metadata
+        return metadata
 
     def iter_blocks_with_metadata(
         self,
     ) -> Iterator[Tuple[ObjectRef[Block], BlockMetadata]]:
+        """Iterate over the blocks along with their metadata.
+
+        The length of this iterator is not known until execution.
+
+        Returns:
+            An iterator of block references and the corresponding block metadata.
+        """
         context = DatasetContext.get_current()
         outer = self
 
         class Iter:
             def __init__(self):
-                self._base_iter = outer._iter_block_partitions()
+                self._base_iter = outer._iter_block_partition_refs()
+                self._pos = -1
                 self._buffer = []
 
             def __iter__(self):
                 return self
 
             def __next__(self):
                 while not self._buffer:
+                    self._pos += 1
                     if context.block_splitting_enabled:
                         part_ref, _ = next(self._base_iter)
                         partition = ray.get(part_ref)
                     else:
-                        block, metadata = next(self._base_iter)
+                        block, _ = next(self._base_iter)
+                        metadata = outer._tasks[self._pos].get_metadata()
```

Comment:
I originally had this force full metadata (by having it block on each submitted read task and fetching the resulting block metadata), but this would be a change in semantics, so to de-risk this PR, I reverted it back to the old semantics. See [this commit](https://github.com/ray-project/ray/pull/23821/commits/acadb9bf84919415f5b5ab4d744b364b758a358b) for the semantics that I previously had.

Reply:
Ok, maybe document the iter_blocks_with_metadata() about what it's returning.